### ISSUE TITLE
Performance problem for defconsts for large object in certified book

### ISSUE THREAD

**Issue by MattKaufmann** [acl2/acl2#502]
Thanks to David Rager for working with me to come up with this Issue.

In short: defconsts (defined in std/util/defconsts.lisp) can be very slow
during include-book, and a possible solution is to have some option for
defconsts that specifies the use of hons-copy on the generated value.

Details:

If defconsts is used to define a large cons structure in two different
books, then after including one book (or evaluating that defconsts in
the loop), ACL2 can go out to lunch when loading the compiled file for
the second book.  The following example illustrates the problem.
First, create two identical books bar.lisp and bar2.lisp, each with
the following contents.

    (in-package "ACL2")
    (include-book "std/util/defconsts" :dir :system)
    (defun make-tree (n)
      (declare (type (integer 0 *) n))
      (if (zp n)
          nil
        (let ((x (make-tree (1- n))))
          (cons x x))))
    (defconsts *a* (hons-copy (make-tree 50)))

You will see ACL2 go out to lunch when you execute the following two
forms, during loading of the compiled file for bar2.

    (include-book "bar")
    (include-book "bar2")

Below I'll explain what I believe is going on.  First, here is some
relevant background.

(1) When include-book is executed, the compiled file for the book (if
it exists) is loaded first, before the events in the book are
executed.

(2) (defconsts NAME FORM) macroexpands to a make-event form that
generates an event (defconst NAME (quote VAL)), where VAL is the value
of FORM.

(3) During loading of a compiled file, when evaluating a defconst
form, the raw-Lisp defconst macro checks whether the proposed defconst
is redundant.  There are two redundancy checks.  The first redundancy
check uses EQUAL to compare the original defconst form for the given
name (such as *a*), if any, with the one provided by the compiled
file.  If that check fails -- for example, consider (defconst *c* 7)
and the redundant (defconst *c* (+ 3 4)) -- then a second redundancy
check is performed, which compares the two values using EQUAL.

(4) For a compiled file, the compilation of a form (defconst NAME
(quote VAL)) does not store any information for honses about VAL.

Now we can analyze the "going out to lunch".  ACL2 starts the second
include-book by loading the compiled file for bar2, by (1).  That
results in raw-Lisp evaluation of the (defmacro *a* (quote val)) where
val is the value of (hons-copy (make-tree 50)), by (2).  Then by (3),
it is checked whether this defmacro form is equal to the corresponding
form from bar1.lisp.  By (4), val is not a hons, and hence this
equality check must be made without the benefit of hons; hence it is
very, very slow.  (I'll also note that the second check, which
compares values rather than event forms, would suffer from the same
problem; so switching the order of these two checks wouldn't help.)

The implementation of (1) is quite complex; in my opinion, it is not a
feasible option to modify DEFCONST to avoid (4).

A workaround is to make the equality checks fast by tweaking the forms
generated by defconsts, as illustrated in the following modification
of the value printed by :trans1 applied to the defconsts form in file
bar1.lisp.  Below, the additional code is shown in lower case, namely:
replace (LIST 'QUOTE A) by the form (list 'identity1 (list 'hons-copy
(LIST 'QUOTE A))).  If we make the same modification to file
bar2.lisp, except naming the identity macro IDENTITY2 instead of
IDENTITY1, then the second include-book is no longer slow.  Why?
Because, as per (3): the first redundancy check immediately fails
since the symbols IDENTITY1 and IDENTITY2 are distinct, and the second
redundancy check immediately succeeds because hons-copy is called
before returning the value to be checked against the existing value of
*a*.  (Here, hons-copy is fast because of the structure sharing.)

    (defmacro identity1 (x) x)
    (MAKE-EVENT
     (TIME$
      (LET ((__FUNCTION__ '|(DEFCONSTS *A* ...)|))
           (DECLARE (IGNORABLE __FUNCTION__))
           (LET ((A (HONS-COPY (MAKE-TREE 50))))
                (LIST 'WITH-OUTPUT
                      :OFF '(EVENT)
                      (LIST 'PROGN
                            (LIST 'DEFCONST
                                  '*A*
                                  (list 'identity1
                                        (list 'hons-copy
                                              (LIST 'QUOTE A))))))))
      :MSG "; ~s0: ~st seconds, ~sa bytes~%"
      :ARGS (LIST "(DEFCONSTS *A* ...)")))

A possible solution, then, is to change defconsts to permit the user
to specify that the value is to be a hons.  For example, the form

    (defconsts *a* (make-tree 50) :hons identity1)

could macroexpand to the two forms displayed just above.

**Comment by jaredcdavis** [comment]
Hi,

This seems like a clever workaround.  If I understand things correctly, I think
this has nothing in particular to do with `defconsts`; we can adjust your
`bar.lisp` and `bar2.lisp` to instead contain the following:

```
    (in-package "ACL2")

    (defun make-tree (n)
      (declare (type (integer 0 *) n))
      (if (zp n)
          nil
        (let ((x (make-tree (1- n))))
          (cons x x))))

    (make-event
     `(defconst *a* ',(hons-copy (make-tree 50))))
```

And then loading both files still goes out to lunch.

That isn't to say that we shouldn't consider doing something to `defconsts` to
help to avoid this.  On the other hand, I guess this strikes me as a kind of
peculiar thing to want to do.  That is, do you really want to be defining the
same large constant in separate files?

Maybe you have good and important reasons to do this, but even if that's the
case, it seems like you're already going to have to know that you need to write
these custom identity functions, and, well, I just wonder whether it's really
worth complicating the interface to `defconsts` to support them?  Would it save
you any work versus writing your `make-event` forms that generate custom
`defconst` forms with the extra identity functions in the right places?

In the serialize code we go to some efforts to remember which constants are
honsed, which alists are fast, etc., so that we can try to preserve this as
necessary.  For instance, if I make a new file, `foo.lisp`:

```
    (in-package "ACL2")
    (include-book "std/util/defconsts" :dir :system)
    (defconsts *obj* (list (hons 1 2) (cons 3 4)))
```

Then the normed status is preserved after loading the file:

```
    (include-book "system/hons-check/hons-check" :dir :system)
    (include-book "foo")
    (unsound-normedp *obj*)           ;; nil
    (unsound-normedp (first *obj*))   ;; t
    (unsound-normedp (second *obj*))  ;; nil
```

I guess not sure how this should work when constants are defined redundantly.
At present it looks like the first book that gets loaded seems to "win".  That
is, if I introduce `foo2.lisp` as follows:

```
    (in-package "ACL2")
    (include-book "std/util/defconsts" :dir :system)
    (defconsts *obj* (list (cons 1 2) (hons 3 4)))
```

Then I find:

```
    (include-book "system/hons-check/hons-check" :dir :system)
    (include-book "foo")
    (include-book "foo2")
    (unsound-normedp *obj*)           ;; nil
    (unsound-normedp (first *obj*))   ;; t
    (unsound-normedp (second *obj*))  ;; nil   -- "foo wins"
```

But if I switch the order of the `include-book`s, I find:

```
    (include-book "system/hons-check/hons-check" :dir :system)
    (include-book "foo2")
    (include-book "foo")
    (unsound-normedp *obj*)           ;; nil
    (unsound-normedp (first *obj*))   ;; nil
    (unsound-normedp (second *obj*))  ;; t     -- "foo2 wins"
```

So I guess we are basically just restoring the honses from the first
certificate that is loaded, and ignoring whatever is honsed in the redundant
defconst later.  I guess that seems pretty sensible.

I don't know much of anything about compiled files, but I sort of wonder if we
might be able to get away with not even storing the actual constant values in
the compiled file at all?  In particular...

  - To support undo and so forth, I thought that each ACL2 `defconst` becomes a
    special variables in Lisp, i.e., probably what goes into the compiled file
    is something like `(defvar *obj* ...)` or `(defparameter *obj* ...)`.

  - We surely must be `setq`'ing constants like `*obj*` to the value we read out of
    the certificate, because otherwise how could we be getting these honses
    restored?

So if that's all correct, then it seems like the actual value we store in the
compiled file can't possibly matter after the `setq`.  Can it matter before then?
Certainly we don't have to worry about, e.g., functions that might include
`*obj*` in their body, because from Lisp's perspective `*obj*` is a variable
instead of a constant, so the compiler has to be ready for `*obj*` to have a
different value later.  Maybe it could matter if the compiled files that ACL2
generates include fancy `eval-when` constructs where the value of `*obj*` might
actually be relevant, but I don't know if they do that or not.

Anyway, probably none of this is helpful, but I think my general feeling is
that we probably shouldn't try to extend `defconsts` to work around this.
Please feel free to push back if this seems unreasonable...

Cheers,
Jared

**Comment by MattKaufmann** [comment]
I agree that avoiding redundant defconst (or defconsts) forms is the
best way to deal with such issues.  So I'm fine with the position that
defconsts isn't to be changed to deal with this.  Also, I agree that
defconst alone is still somewhat of an issue.  At first I thought not
-- I had in mind that defconsts might generate a form that differs a
bit from yours, Jared, by putting hons-copy on the outside:

    (make-event
     `(defconst *a* (hons-copy ',(make-tree 50))))

But I still run into problems here.  I can avoid the "out-to-lunch"
problem for (defconst VAR TERM) during loading of the compiled file,
by modifying the raw-Lisp definition of defconst to avoid the equality
check of event forms when TERM has more than 200,000 conses.  But then
ACL2 function defconst-fn goes out to lunch when checking redundancy
of the event form.  Sigh.

I've made a note to consider fixing that, but I prefer not to track
this here, and it seems that avoiding redundant large defconst (and
defconsts) forms is the best way to go anyhow.  So I'm closing this
issue.

Oh, and here is a final note, about your suggestion Jared of possibly
"not even storing the actual constant values in the compiled file at
all".  The problem with that is macros, illustrated by evaluating the
following in raw Lisp (to mimic what would be done during loading a
compiled file).

    ; Avoid the following intended value:
    ; (defconst *a* 'imagine-big-value-here)

    ; Kind of "stub out" *a*.
    (defconst *a* nil)

    (defmacro foo () (list 'quote *a*))

    (defun bar () (foo))

    ; Returns NIL instead of IMAGINE-BIG-VALUE-HERE.
    (bar)

Even were this not a problem, I'd loathe messing with the compiled
file.  The situation there is quite complicated, much more than just
generating a defparameter, because of early load of compiled files
(storing definitions temporarily in hash tables), property
'redundant-raw-lisp-discriminator, and probably other things I'm not
even thinking of.

**Comment by ragerdl** [comment]
Thank `y'all` for considering this.  I'll mention that the reason I avoided putting this shared form in a shared book is that it's now the only form in that book (it doesn't belong in the book that reads in the verilog imho for another reason).  Often in the work I'm doing, adding an extra book on the critical path adds a measurable amount of time to the build time.  After I thought about it, I realized that this was why I was avoiding the extra book.

I know the community has done a lot to speedup `include-book`, but I still try to avoid `include-book` when there isn't a clear reason for it.  I suspect that this particular case isn't too bad (somewhere between 1 and 90 extra seconds), and there's a concrete reason for having the extra layer of indirection.  So, I'm fine with the resolution.
